graphdatascience/graph/graph_cypher_runner.py

```diff
@@ -0,0 +1,135 @@
+
+
+class GraphCypherRunner(CallerBase):
+    def __init__(self, query_runner: QueryRunner, namespace: str, server_version: ServerVersion) -> None:
+        if server_version < ServerVersion(2, 4, 0):
+            raise ValueError("The new Cypher projection is only supported since GDS 2.4.0.")
+        super().__init__(query_runner, namespace, server_version)
+
+    def project(
+        self,
+        query: str,
+        database: Optional[str] = None,
+        **params: Any,
+    ) -> GraphCreateResult:
+        """
+        Run a Cypher projection.
+        The provided query must end with a `RETURN gds.graph.project(...)` call.
+
+        Parameters
+        ----------
+        query: str
+            the Cypher projection query
+        database: str
+            the database on which to run the query
+        params: Dict[str, Any]
+            parameters to the query
+
+        Returns
+        -------
+        A tuple of the projected graph and statistics about the projection
+        """
+
+        GraphCypherRunner._verify_query_ends_with_return_clause(self._namespace, query)
+
+        # See run_cypher
+        qr = self._query_runner
+
+        # The Arrow query runner should not be used to execute arbitrary Cypher
+        if isinstance(qr, ArrowQueryRunner):
+            qr = qr.fallback_query_runner()
+
+        result = qr.run_query(query, params, database, False)
+        result = result.squeeze()
+        try:
+            graph_name = str(result["graphName"])
+        except (KeyError, TypeError):
+            if isinstance(result, str):
+                if result == str(self._server_version):
+                    graph_name = (
+                        "Could not get the graph name from the result. "
+                        "This is probably because this is a unit test and no mock result was given to the query runner"
+                    )
+                else:
+                    # This is likely a test where a mock result was provided
+                    # But it was a dict with a single entry.
+                    # squeeze() will have removed the dict and only left that single value
+                    # so we assume that the key was graphName and use the value.
+                    graph_name = result
+            else:
+                raise ValueError(
+                    f"Invalid query, the query must end with the `RETURN {self._namespace}(...)` call: {query}"
+                )
+
+        return GraphCreateResult(Graph(graph_name, self._query_runner, self._server_version), result)  # type: ignore
+
+    __separators = re.compile(r"[,(.]")
+
+    @staticmethod
+    def _verify_query_ends_with_return_clause(namespace: str, query: str) -> None:
+        """
+        Verifies that the query ends in a `RETURN gds.graph.project(...)` call.
+        Invalid queries will raise a ValueError.
+        """
+
+        # We iterate through the "tokens" of the query, where a token here is
```

Comment:
it looks like a regexp could also work here, but I also think your version could be more maintainable :)

Reply:
This reminds me of https://stackoverflow.com/a/1732454 ^^